src/net.cpp

```diff
@@ -2571,7 +2587,8 @@ void CConnman::DeleteNode(CNode* pnode)
     assert(pnode);
     bool fUpdateConnectionTime = false;
     m_msgproc->FinalizeNode(pnode->GetId(), fUpdateConnectionTime);
-    if (fUpdateConnectionTime) {
+    if (fUpdateConnectionTime && !pnode->IsBlockOnlyConn()) {
```

Comment:
Agreed that this is confusing.  Originally I was contemplating have code in `net.cpp` that wouldn't call `CAddrMan::Attempt()` on block-relay-only peers, and so having this logic also be in `net.cpp` made more sense to me; but now I think it's fine to move into `net_processing`/`FinalizeNode()`.

Reply:
I'd suggest going further and removing `fCurrentlyConnected`, and then changing `FinalizeNode()` to check whether the peer is:
- `fSuccessfullyConnected`
- not inbound and not block relay only
- has 0 misbehavior score
and then directly call `CAddrMan.Connected()` directly, so that all the logic around calling `Connected()` is in one place.

If you agree, perhaps you could take a look at 1755f08b4dba34b483a401c6098d5df549219ce7 in PR #20228, which does some of that. For this PR, just putting the block relay only check inside `FinalizeNode()` at least doesn't make the situation any worse.

Reply:
I just did the simple thing of moving this to `FinalizeNode()`.  Because there's no good way to get a `CNode` from a `NodeId` while it's in the process of being deleted (as far as I can tell!), I had to change the interface to accommodate this move.

Reply:
`DeleteNode()` is holding a reference to the `CNode`, so you should just be able to pass a const reference to that CNode down to `FinalizeNode()`. That'd be more symmetrical with the other `NetEventsInterface` interface functions (you could also pass a pointer like the other functions, but they should all be const references really)